src/hdmf/build/classgenerator.py

```diff
@@ -0,0 +1,344 @@
+        parent_docval_args = set(arg['name'] for arg in get_docval(base.__init__))
+        new_args = list()
+        for attr_name, field_spec in not_inherited_fields.items():
+            # auto-initialize arguments not found in superclass
+            if attr_name not in parent_docval_args:
+                new_args.append(attr_name)
+
+        @docval(*docval_args)
+        def __init__(self, **kwargs):
+            if name is not None:  # force container name to be the fixed name in the spec
+                kwargs.update(name=name)
+            pargs, pkwargs = fmt_docval_args(base.__init__, kwargs)
+            base.__init__(self, *pargs, **pkwargs)  # special case: need to pass self to __init__
+
+            for f in new_args:
+                arg_val = kwargs.get(f, None)
+                setattr(self, f, arg_val)
+        classdict['__init__'] = __init__
+
+
+class TypeDoesNotExistError(Exception):  # pragma: no cover
+    pass
+
+
+class CustomClassGenerator:
+    """Subclass this class and register an instance to alter how classes are auto-generated."""
+
+    def __new__(cls, *args, **kwargs):  # pragma: no cover
+        raise TypeError('Cannot instantiate class %s' % cls.__name__)
+
+    # mapping from spec types to allowable python types for docval for fields during dynamic class generation
+    # e.g., if a dataset/attribute spec has dtype int32, then get_class should generate a docval for the class'
+    # __init__ method that allows the types (int, np.int32, np.int64) for the corresponding field.
+    # passing an np.int16 would raise a docval error.
+    # passing an int64 to __init__ would result in the field storing the value as an int64 (and subsequently written
+    # as an int64). no upconversion or downconversion happens as a result of this map
+    _spec_dtype_map = {
+        'float32': (float, np.float32, np.float64),
+        'float': (float, np.float32, np.float64),
+        'float64': (float, np.float64),
+        'double': (float, np.float64),
+        'int8': (np.int8, np.int16, np.int32, np.int64, int),
+        'int16': (np.int16, np.int32, np.int64, int),
+        'short': (np.int16, np.int32, np.int64, int),
+        'int32': (int, np.int32, np.int64),
+        'int': (int, np.int32, np.int64),
+        'int64': np.int64,
+        'long': np.int64,
+        'uint8': (np.uint8, np.uint16, np.uint32, np.uint64),
+        'uint16': (np.uint16, np.uint32, np.uint64),
+        'uint32': (np.uint32, np.uint64),
+        'uint64': np.uint64,
+        'numeric': (float, np.float32, np.float64, np.int8, np.int16, np.int32, np.int64, int, np.uint8, np.uint16,
+                    np.uint32, np.uint64),
+        'text': str,
+        'utf': str,
+        'utf8': str,
+        'utf-8': str,
+        'ascii': bytes,
+        'bytes': bytes,
+        'bool': (bool, np.bool_),
+        'isodatetime': datetime,
+        'datetime': datetime
+    }
+
+    @classmethod
+    def _get_scalar_type_map(cls, spec_dtype):
```

Comment:
This function name is kind of confusing. Is it exposed for overwriting?

Reply:
This used to be in `TypeMap.__get_scalar_type_map` so we can feel free to change it. I made it single underscore in case a subclass of `CustomClassGenerator` wants to customize it but maybe that is not necessary.

Reply:
How about renaming it to `_get_type_from_spec_dtype`

Reply:
I think it's fine for now. as long as it's not out in the wild and it's documented, I think it's fine.